[AArch64] Remove superfluous sxtw in peephole opt

davemgreen opened the pull request:
Across a basic-block we might have in i32 extract from a value that only operates on upper bits (for example a sxtw). We can replace the COPY with a new version skipping the sxtw.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp:
```diff
@@ -690,6 +691,33 @@ bool AArch64MIPeepholeOpt::visitFMOVDr(MachineInstr &MI) {
+
+  MachineInstr *SrcMI = MRI->getUniqueVRegDef(MI.getOperand(1).getReg());
+  MachineInstr *CopyMI = SrcMI;
+  if (SrcMI && SrcMI->isFullCopy() &&
+      MRI->hasOneNonDBGUse(SrcMI->getOperand(1).getReg()))
+    SrcMI = MRI->getUniqueVRegDef(SrcMI->getOperand(1).getReg());
```
Could there be more copies in the chain (i.e. `SrcMI` could be a COPY after line 706). A while loop would eventually get to the end of a long copy chain.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64MIPeepholeOpt.cpp:
```diff
@@ -690,6 +691,33 @@ bool AArch64MIPeepholeOpt::visitFMOVDr(MachineInstr &MI) {
+// operates on upper bits (for example a sxtw). We can replace the COPY with a
+// new version skipping the sxtw.
+bool AArch64MIPeepholeOpt::visitCopy(MachineInstr &MI) {
+  if (MI.getOperand(1).getSubReg() != AArch64::sub_32 ||
+      !MRI->hasOneNonDBGUse(MI.getOperand(1).getReg()))
+    return false;
+
+  MachineInstr *SrcMI = MRI->getUniqueVRegDef(MI.getOperand(1).getReg());
```
A minor improvement is storing `MI->getOperand(1)` to a variable rather than calling `getOperand()` a few times.

davemgreen committed:
[AArch64] Remove superfluous sxtw in peephole opt

Across a basic-block we might have in i32 extract from a value that only
operates on upper bits (for example a sxtw). We can replace the COPY with a
new version skipping the sxtw.

davemgreen commented:
Ping - thanks.

SamTebbs33 reviewed (approved):
Thanks for the improvements and sorry for missing your update, looks great to me.

krasimirgg commented:
I rolled back the patch yesterday, sorry for not leaving a comment about that here in a timely manner.

We saw some issues with this in some halide-generated code.

It appears that after this, llc segfaults on the following code:
```
target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32"
target triple = "aarch64--linux-eabi"

define i32 @f(i32 %0) {
entry:
  %1 = sext i32 %0 to i64
  br label %A

A:
  %2 = trunc i64 %1 to i32
  %a69.us = sub i32 0, %2
  %a69.us.fr = freeze i32 %a69.us
  %3 = zext i32 %a69.us.fr to i64
  br label %B

B:
  %t = icmp eq i64 0, %3
  br i1 %t, label %A, label %B
}
```

`assert.h assertion failed at .../llvm/lib/CodeGen/LiveVariables.cpp:159 in void llvm::LiveVariables::HandleVirtRegUse(Register, MachineBasicBlock *, MachineInstr &): MRI->getVRegDef(Reg) && "Register use before def!"`

davemgreen commented:
Thanks - the reproducer in the commit message was useful for figuring out what was wrong. It just needs to make sure it removed all the copies. I'll put together a recommit - please let me know if you run into any more problems, and we can revert again to make sure the branch is clean.
